dbt_meshify/change.py

```diff
@@ -0,0 +1,100 @@
+import dataclasses
+from enum import Enum
+from pathlib import Path
+from typing import Dict, Iterable, List, Optional, Protocol
+
+
+class Operation(str, Enum):
+    """An operation describes the type of work being performed."""
+
+    Add = "add"
+    Update = "update"
+    Remove = "remove"
+    Copy = "copy"
+    Move = "move"
+
+
+class EntityType(str, Enum):
+    """An EntityType represents the type of entity being operated on in a Change"""
+
+    Model = "model"
+    Analysis = "analysis"
+    Test = "test"
+    Snapshot = "snapshot"
+    Operation = "operation"
+    Seed = "seed"
+    RPCCall = "rpc"
+    SqlOperation = "sql_operation"
+    Documentation = "doc"
+    Source = "source"
+    Macro = "macro"
+    Exposure = "exposure"
+    Metric = "metric"
+    Group = "group"
+    SemanticModel = "semantic_model"
+    Project = "project"
+    Code = "code"
+
+    def pluralize(self) -> str:
+        if self is self.Analysis:
+            return "analyses"
+        return f"{self.value}s"
+
+
+class Change(Protocol):
+    """A change represents a unit of work that should be performed within a dbt project."""
+
+    operation: Operation
+    entity_type: EntityType
+    identifier: str
+    path: Path
+
+
+@dataclasses.dataclass
+class BaseChange:
+    operation: Operation
+    entity_type: EntityType
+    identifier: str
+    path: Path
+
+
+@dataclasses.dataclass
+class ResourceChange(BaseChange):
+    """A ResourceChange represents a unit of work that should be performed on a Resource in a dbt project."""
+
+    data: Dict
+    source_name: Optional[str] = None
+
+
+@dataclasses.dataclass
+class FileChange(BaseChange):
+    """A ResourceChange represents a unit of work that should be performed on a File in a dbt project."""
+
+    data: Optional[str] = None
+    source: Optional[Path] = None
```

Comment:
any reason to have `ResourceChange.source_name()` and `FileChange.source()`?

Reply:
Yes! Source as a Path vs the string name of a dbt source. I may need better docs.